iceoryx_posh/source/roudi/port_manager.cpp

```diff
@@ -365,38 +445,84 @@ void PortManager::sendToAllMatchingReceiverPorts(const capro::CaproMessage& mess
     }
 }
 
-void PortManager::sendToAllMatchingInterfacePorts(const capro::CaproMessage& message)
+bool PortManager::sendToAllMatchingPublisherPorts(const capro::CaproMessage& message,
+                                                  SubscriberPortProducerType& subscriberSource) noexcept
 {
-    for (auto interfacePortData : m_portPool->interfacePortDataList())
+    bool publisherFound = false;
+    for (auto publisherPortData : m_portPool->getPublisherPortDataList())
     {
-        iox::popo::InterfacePort interfacePort(interfacePortData);
-        // not to the interface the port is located
-        if (message.m_serviceDescription.getSourceInterface()
-            != interfacePort.getCaProServiceDescription().getSourceInterface())
+        PublisherPortRouDiType publisherPort(publisherPortData);
+        if (subscriberSource.getCaProServiceDescription() == publisherPort.getCaProServiceDescription())
         {
-            interfacePort.dispatchCaProMessage(message);
+            auto publisherResponse = publisherPort.dispatchCaProMessage(message);
+            if (publisherResponse.has_value())
+            {
+                // send response to subscriber port
+                auto returnMessage = subscriberSource.dispatchCaProMessage(publisherResponse.value());
+
+                // ACK or NACK are sent back to the subscriber port, no further response from this one expected
+                cxx::Ensures(!returnMessage.has_value());
+
+                // inform introspection
+                m_portIntrospection.reportMessage(publisherResponse.value());
+            }
+            publisherFound = true;
         }
     }
+    return publisherFound;
 }
 
-bool PortManager::areAllReceiverPortsSubscribed(const ProcessName_t& processName)
+void PortManager::sendToAllMatchingSubscriberPorts(const capro::CaproMessage& message,
+                                                   PublisherPortRouDiType& publisherSource) noexcept
 {
-    int32_t numberOfReceiverPorts{0};
-    int32_t numberOfConnectedReceiverPorts{0};
-    for (auto receiverPortData : m_portPool->receiverPortDataList())
+    for (auto subscriberPortData : m_portPool->getSubscriberPortDataList())
     {
-        ReceiverPortType receiver(receiverPortData);
-        if (receiver.getProcessName() == processName)
+        SubscriberPortProducerType subscriberPort(subscriberPortData);
+        if (subscriberPort.getCaProServiceDescription() == publisherSource.getCaProServiceDescription())
```

Comment:
somehow I always think this checks if getting the service description is equal, not the service description itself. That's why I dislike the `get` prefix 😉